Userland/Utilities/crypto.cpp

```diff
@@ -0,0 +1,239 @@
+    fill_with_random(iv);
+
+    auto nonce = TRY(ByteBuffer::create_uninitialized(12));
+    fill_with_random(nonce);
+
+    run_benchmark_with_all_sizes(name, [&](auto& buffer) {
+        Optional<Algorithm> cipher;
+        if constexpr (needs_nonce)
+            cipher = Algorithm(key.bytes(), nonce);
+        else
+            cipher = Algorithm(key.bytes(), key_bits, Crypto::Cipher::Intent::Encryption);
+        auto out_bytes = out_buffer.bytes();
+        if constexpr (requires { cipher->encrypt(buffer, out_bytes, iv.bytes()); })
+            cipher->encrypt(buffer, out_bytes, iv.bytes());
+        else
+            cipher->encrypt(buffer, out_bytes);
+        use(out_buffer);
+    });
+    return {};
+}
+
+static ErrorOr<void> benchmark(StringView algorithm)
+{
+#define BENCH(name, type, algo, ...)                                                       \
+    if (algorithm == #name) {                                                              \
+        outln("Benchmarking {}...", #name);                                                \
+        return run_##type##_benchmark<Crypto::algo>(#name##sv __VA_OPT__(, ) __VA_ARGS__); \
+    }
+
+    ALL_ALGORITHMS(BENCH);
+
+#undef BENCH
+
+    return Error::from_string_literal("Unknown algorithm");
+}
+
+void dump_benchmark_results()
+{
+    // algo, size, min, max, avg, throughput
+    outln("{:<20} {:<10} {:<10} {:<10} {:<10} {:<10}", "Algorithm", "Size", "Min us/op", "Max us/op", "Avg us/op", "Throughput");
+    for (auto& [algo, timings] : g_all_timings) {
+        for (auto size : sizes_in_bytes) {
+            auto t = timings.get(size);
+            if (!t.has_value())
+                continue;
+            auto& timing = t.value();
+            outln("{:<20} {:<10} {:<10} {:<10} {:<10} {:<10}/s",
+                algo,
+                human_readable_size(timing.unit_bytes),
+                timing.min_us,
+                timing.max_us,
+                timing.total_us / timing.count,
+                human_readable_quantity(timing.unit_bytes * timing.count / TimeSlicePerSize.to_microseconds() * 1'000'000));
+        }
+    }
+}
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    StringView command; // [benchmark, run]
+    StringView algorithm;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(command, "Command to run", "command", Core::ArgsParser::Required::Yes);
+    args_parser.add_option(algorithm, "Algorithm for the operation", "algorithm", 'a', "algorithm");
+
+    args_parser.parse(arguments);
+
+    if (command == "benchmark"sv) {
+        auto found = false;
+#define BENCH(name, type, ...)     \
+    if (algorithm == #type) {      \
+        TRY(benchmark(#name##sv)); \
+        found = true;              \
+    }
+
+        ALL_ALGORITHMS(BENCH);
+#undef BENCH
+
+        if (!found)
+            TRY(benchmark(algorithm));
+        dump_benchmark_results();
+    } else if (command == "run"sv) {
+        // run(algorithm);
```

Comment:
why is this commented out? probably just toss this entire branch, and remove `run` from the "commands" list above

Reply:
:soon: